### ISSUE TITLE
SNMP MIB objects MAX-ACCESS respected?

### ISSUE THREAD

**Issue by dsecca-mss** [etingof/pysnmp#325]
I am having an issue where performing an SNMPWalk returns Mib symbols that are meant to be non accessible. I am using pySNMP v4.4.9.

In a custom MIB, I have defined a table and declared the row index and row status columns with MAX-ACCESS as not-accessible:

```
-- The ADAS Error Table
adasErrorTable OBJECT-TYPE
    SYNTAX SEQUENCE OF AdasErrorEntry
    MAX-ACCESS not-accessible
    STATUS current
    DESCRIPTION
        "A list of ADAS Errors"
    ::= {adasSystemTables 1}
    
adasErrorEntry OBJECT-TYPE
    SYNTAX AdasErrorEntry
    MAX-ACCESS not-accessible
    STATUS current
    DESCRIPTION 
        "A table row representing an error"
    INDEX {adasErrorEntryHashID}
    ::= {adasErrorTable 1}
    
AdasErrorEntry ::= SEQUENCE {
    adasErrorEntryHashID Unsigned32,
    adasErrorCode Unsigned32,
    adasErrorDesc DisplayString,
    adasErrorRowStatus RowStatus
}

adasErrorEntryHashID OBJECT-TYPE
    SYNTAX Unsigned32
    MAX-ACCESS _**not-accessible**_
    STATUS current
    DESCRIPTION
        "Adas error entry ID"
    ::= { adasErrorEntry 1}

adasErrorCode OBJECT-TYPE
    SYNTAX Unsigned32
    MAX-ACCESS read-only
    STATUS current
    DESCRIPTION
        "Adas error code"
    ::= { adasErrorEntry 2}

adasErrorDesc OBJECT-TYPE
    SYNTAX DisplayString
    MAX-ACCESS read-only
    STATUS current
    DESCRIPTION
        "Description of ADAS Error code"
    ::= { adasErrorEntry 3}

adasErrorRowStatus OBJECT-TYPE
    SYNTAX RowStatus
    MAX-ACCESS _**not-accessible**_
    STATUS current
    DESCRIPTION
        "This OID uses six main statuses:
        active(1)         is in use and available in table
        notinService(2)   it is present but not yet created
        notReady(3)       it is present but missing info
        createAndGo(4)    create entry in table.  Row will be
                          added to this table if necessary.
        createAndWait(5)  add row to this table
        destroy(6)        will remove the row
    ::= { adasErrorEntry 4}
```

In debugging, I noticed that the adasErrorRowStatus object in the mibSymbols dictionary of the mibBuilder shows a max access of “readonly” even though the Mib I pasted above has it as not-accessible.

In the SNMPv2-SMI.py file, the MibTableColumn class inherits from MibScalar and in MibScalar, I notice that the max-access is hardcoded to “readonly”

The compiled Mib does not show the row status or row index objects calling setMaxAccess, so it looks like the max-access value just defaults to “readonly”
 
Is there a way to force these values as not-accessible at the Mib compile time or do I have to manually change the compiled Mib Python file to add calls to setMaxAccess and pass 'not-accessible'? Would it be possible to change the maxAccess class member of MibScalar to 'not-accessible' rather than 'readonly'?

Thanks for your help!

**Comment by etingof** [comment]
What happens if you just compile your MIB by hand with [mibdump.py](http://snmplabs.com/pysmi/mibdump.html) tool?

```
mibdump.py --rebuild --generate-mib-texts YOUR-MIB
```

The output goes to ~/.pysnmp/mibs` by default. We should have your MAX-ACCESS setting in the compiled MIB. If it is not there, that might be pysmi's problem...

**Comment by dsecca-mss** [comment]
Thank you for your quick response, Ilya!

I ran the mibdump.py script like so:
`mibdump.py --mib-source . --mib-source http://mibs.snmplabs.com/asn1/@mib@ --destination-directory . --destination-format pysnmp MY-MIB.txt`

I still do not see the setMaxAccess call for the table index and RowStatus columns. For the table in my initial comment above, this is what I see in the compiled python file for the table index and row status:

`adasTestID = MibTableColumn((1, 3, 6, 1, 4, 1, 60000, 1, 1, 1, 2, 2, 1, 1), Unsigned32())`
`adasTestRowStatus = MibTableColumn((1, 3, 6, 1, 4, 1, 60000, 1, 1, 1, 2, 2, 1, 5), RowStatus())`

whereas for other table columns, in which I have defined a max-access of 'read-only,' I can see the setMaxAccess call setting it to 'read-only.' For example, one of my columns is:

`adasTestDesc = MibTableColumn((1, 3, 6, 1, 4, 1, 60000, 1, 1, 1, 2, 2, 1, 2), DisplayString()).setMaxAccess("readonly")`

I also still see that the max-access is read-only in the objects stored in mibSymbols from the mibBuilder object in my Agent.

**Comment by etingof** [comment]
Which `pysmi` version are you using? If not the latest, it makes sense to upgrade.

Alternatively, if you can pull [pysmi master](https://github.com/etingof/pysmi) and run the same  `mibdump.py` command, we'd see if this problem has been fixed in the yet-unreleased pysmi.

**Comment by dsecca-mss** [comment]
I'm using version 0.3.4

I just tried compiling using the mibdump.py from the master branch of the pysmi repo and I get the same output. No setMaxAccess calls for objects needing to be not accessible.

**Comment by etingof** [comment]
My test with pysmi master seems to succeeded. I've taken IF-MIB as a guinea pig and changed `ifStackStatus` object's MAX-ACCESS to not-accessible. Then mibdump produced this code:

```

_IfStackStatus_Type = RowStatus
_IfStackStatus_Object = MibTableColumn
ifStackStatus = _IfStackStatus_Object(
    (1, 3, 6, 1, 2, 1, 31, 1, 2, 1, 3),
    _IfStackStatus_Type()
)
ifStackStatus.setMaxAccess("not-accessible")
if mibBuilder.loadTexts:
    ifStackStatus.setStatus("current")
```

So I think pysmi 0.4.0 (which is master) should work just fine in this respect. My only concern is if it would require bumping pysnmp to master version as well. May be not...

**Comment by dsecca-mss** [comment]
Hmmm interesting. I will try it again from my end, while upgrading to pysnmp since i have not tried this yet. I will let you know what happens

**Comment by etingof** [comment]
> Hmmm interesting. I will try it again from my end, while upgrading to pysnmp since i have not tried this yet. I will let you know what happens

Just keep in mind that pysnmp is loosely dependent on pysmi and vice versa. I'd first make sure that pysmi compiles your MIB into something reasonable. Then I'd try to feed compiled MIB to released pysnmp to see if it would work.

I think it should for the client side scenario (SNMP manager), I don't think it would work for server side MIB implementation (SNMP agent).

The last resort would be to upgrade to the latest pysnmp. Or you can come up with your own Jinja2 [template](https://github.com/etingof/pysmi/blob/master/pysmi/codegen/templates/pysnmp/mib-definitions.j2) that would produce compiled MIB which is compatible with whatever pysnmp version you are at.

**Comment by dsecca-mss** [comment]
So trying again with pysmi master yields the same thing for my MIB. I also tried doing the same test you did with the IF-MIB and I also don't get the not-accessible set by setMaxAccess for ifStackStatus. I've attached the compiled file for IF-MIB if you'd like to take a look at it:
[IF-MIB.zip](https://github.com/etingof/pysnmp/files/3841554/IF-MIB.zip)
I'm compiling using the same command I used before:
`mibdump.py --mib-source . --mib-source http://mibs.snmplabs.com/asn1/@mib@ --destination-directory . --destination-format pysnmp MY-MIB.txt`

Is that Jinja2 template new? I don't see it in pysmi 0.3.4 (The version I have)

**Comment by etingof** [comment]
You seem to be using pysmi 0.3.4 or the file is not overwritten or it's some other file. ;-)

```
# PySNMP MIB module IF-MIB (http://snmplabs.com/pysmi)
# ASN.1 source file://.\IF-MIB.txt
# Produced by pysmi-0.3.4 at Wed Nov 13 09:09:16 2019
# On host ? platform ? version ? by user ?
# Using Python version 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 21:26:53) [MSC v.1916 32 bit (Intel)]
```

**Comment by dsecca-mss** [comment]
haha I'm pretty sure I'm running the correct version of mibdump.py since I'm running it directly in my local folder of the pysmi repo. I confirm that I have the master branch checked out as well. Are you sure you don't have pysmi version 0.3.4 in the comments when you compile IF-MIB on your side? I tried changing ifStackStatus from read-only to not-accessible and back, and I see that the Python file is being updated...

**Comment by etingof** [comment]
May be your python is importing something unexpected. Here's my compiled file:

```
ASN.1 source file://./IF-MIB
Produced by pysmi-0.4.0 at Wed Nov 13 14:12:21 2019
On host igarlic platform Darwin version 18.7.0 by user ilya
Using Python version 3.7.1 (v3.7.1:260ec2c36a, Oct 20 2018, 03:13:28) 
"""
```

**Comment by dsecca-mss** [comment]
Yes, you're right! My Python environment kept on pointing to version 0.3.4. I've installed pysmi directly from the master branch and my MIB is compiling correctly with the proper accessibility.

Thank you very much for your help, Ilya!

Do you know when version 0.4.0 of pysmi will be officially released?

**Comment by etingof** [comment]
> Do you know when version 0.4.0 of pysmi will be officially released?

That's on my TODO list. I need some time to wrap up things with the new major release.

Have you tried the new MIB with old pysnmp?

**Comment by dsecca-mss** [comment]
> Have you tried the new MIB with old pysnmp?

Yes, I compiled the MY-MIB.txt with pysmi 0.4.0 and then stepped into my agent which uses pysnmp 4.4.9 and I see the MIB objects containing the correct accessibility within mibBuilder. I will need to test things further to ensure that nothing else gets broken in my agent but for now, it looks good.

**Comment by etingof** [comment]
Awesome! Is it some SNMP manager application you are working on? I bet the new MIB won't work for SNMP agent based on pysnmp 4...

**Comment by dsecca-mss** [comment]
Yes, I have both a Manager and an Agent that I developed. The agent seems to be working fine however, I have not tested the Manager yet.
